Added Normaliz as a linked library

mahrud opened the pull request:
- **added support for torsion class group for truncate**
- **renamed truncate option Nef to Cone**
- **fixed caching in truncationMonomials**
- **simplified hilbertBasis Cone in Polyhedra**
- **modified isVeryAmple ToricDivisor to use hilbertBasis Cone**
- **modified latticePoints ToricDivisor to use latticePoints Cone**
- **added Nauty as a library in CMake**
- **added Normaliz as a library in CMake**
- **updated installation documents and docker scripts**
- **added rawHilbertBasis using libnormaliz**
- **modified truncate, hilbertBasis Cone to use rawHilbertBasis**
- **moved hilbertBasis to Polyhedra**
- **fixed broken tests due to new order of hilbertBasis**
- **fixed a dependency in ReactionNetworks**
- **fixed a dependency in a GroebnerStrata example**
- **added rawFourierMotzkin, calling libnormaliz**
- **changed findHeft to use rawFourierMotzkin**
- **brushed up Normaliz, in preparation for overhaul**
- **moved isSmooth to Core**
- **moved isVeryAmple to Core**
- **fixed a conflict over isReflexive**
- **removed pesky printout**
- **added libnauty2-dev and libnormaliz-dev to github builds**
- **fixed broken examples in FourierMotzkin**
- **fixed symbol conflicts in Normaliz examples**
- **fixed bug in heft of degree zero rings**
- **fixed symbol conflicts in MonodromySolver examples**
- **wip: fix configure.ac for libnormaliz**

@d-torrance the last commit needs work for getting autotools to find and link with Normaliz. Could you help with this?

d-torrance commented:
Will do!

d-torrance commented:
Just pushed a commit updating the autotools build.  It's not quite working yet when we're building normaliz, but it should work when it already exists on the system.

mahrud commented:
Interesting. We don't need antic for now, but seems like Nauty is needed also:
```
Undefined symbols for architecture x86_64:
  "_densenauty", referenced from:
      libnormaliz::nauty_result<long> libnormaliz::compute_automs_by_nauty_Gens_LF<long>(libnormaliz::Matrix<long> const&, ...
```
When building it ourselves we certainly don't need either.

mahrud commented:
That seems to have done the trick. At some point making sure that libnormaliz builds successfully would be good, but I think this is good for now.

mahrud commented on M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2:
```diff
@@ -531,10 +528,7 @@ vertices ToricDivisor := Matrix => D -> (
-    V = transpose (normaliz (transpose V,"polytope"))#"gen";
-    s := select (numColumns V, i -> V_(d,i) === 1);
-    c := (V_s)^{0..d-1};
+    c := matrix(vector \ latticePoints convexHull V);
```
@ggsmith with this PR, `isVeryAmple` and `latticePoints` for toric divisors are instead using `rawHilbertBasis` (called by Polyhedra after some hoops). Other changes in `NormalToricVarieties` are superficial (e.g. the symbol `isSmooth` is moving to Core, resulting in a bunch of documentation changes).

ggsmith commented:
I have a few questions:

- In ```FourierMotzkin```: I don't understand some of the changes the documentation node "Finding the facets of the cyclic polytope".  Why compute the vertices of the cyclic polytope, rather than just constructing them explicitly?  The original code seems more accessible to me.  Perhaps the best option is to include both by verifying that the more elementary construction coincides with the one obtain from the ```cyclicPolytope``` construction.
- In ```NormalToricVarieties```: as I understand it, the methods ```isVeryAmple``` and ```latticePoints``` now both depend on the ```Polyhedra``` package.  The earlier code was explicitly designed to avoid this.  Can this code not be altered to call ```rawHilbertBasis``` directly?  If not, then I would much prefer to wait to make this change until all the Normalize routines are available.
- In ```NormalToricVarieties```: in the documentation nodes where you removed links to the ```(isSmooth, NormalToricVariety)``` page, did you check that this link already appears in the ```SeeAlso``` section?
- In ```NormalToricVarieties/Tests.m2```: Why did you change the ```sort``` to a ```set```?

mahrud commented:
> * In `FourierMotzkin`: I don't understand some of the changes the documentation node "Finding the facets of the cyclic polytope".  Why compute the vertices of the cyclic polytope, rather than just constructing them explicitly?  The original code seems more accessible to me.  Perhaps the best option is to include both by verifying that the more elementary construction coincides with the one obtain from the `cyclicPolytope` construction.

I changed this because of the conflict in redefining `cyclicPolytope` from `Polyhedra`, but I'm happy to expand the documentation to show the explicit construction.

> * In `NormalToricVarieties`: as I understand it, the methods `isVeryAmple` and `latticePoints` now both depend on the `Polyhedra` package.  The earlier code was explicitly designed to avoid this.  Can this code not be altered to call `rawHilbertBasis` directly?  If not, then I would much prefer to wait to make this change until all the Normalize routines are available.

The Polyhedra methods `latticePoints` and `hilbertBasis` are a little obfuscated but are not doing anything fancy, only post-processing the output of `rawHilbertBasis`. Yes, the toric variety methods `isVeryAmple` and `latticePoints` can be altered to call `rawHilbertBasis` directly, but I'm against duplicating the raw post-processing code in Polyhedra. Specifically, computing lattice points of a polytope using a Hilbert basis computation is a piece of code that I think should be defined in one efficient routine rather than several places.

I should add that with this PR, NormalToricVarieties doesn't import the package Normaliz anymore, so I don't understand what would change when all Normaliz routines are available. Do you have another suggestion?

> * In `NormalToricVarieties`: in the documentation nodes where you removed links to the `(isSmooth, NormalToricVariety)` page, did you check that this link already appears in the `SeeAlso` section?

I did not, but instead I changed them all to link to the correct node.

> * In `NormalToricVarieties/Tests.m2`: Why did you change the `sort` to a `set`?

I'm actually not sure why I changed it so drastically, I'll change it back, but importantly `rawHilbertBasis` from Normaliz and 4ti2 have different output orders, which actually results in a desingularization with different order of rays in this example, so in the maximal cones 3 and 5 will be swapped.

mahrud commented:
@jkyang92 could you review the parts of this PR in the engine? I'm not confident if I dealt with memory matters safely.

@mikestillman could you enable requesting reviews from anyone? Currently I can't ask Jay to leave a review because it's only limited to [this team](https://github.com/orgs/Macaulay2/teams/macaulay2-contributors).

mikestillman commented:
@mahrud Do you know how I enable requesting reviews from anyone?  I don't immediately see where to do so...

mahrud commented:
I think `Settings > Moderation Options`.

mikestillman commented:
@mahrud, on github repo settings for M2, I look at 
```
Restrict users who are permitted to approve or request changes on pull requests in this repository.

Limit to users explicitly granted read or higher access
[]When enabled, only users explicitly granted access to this repository will be able to submit pull request reviews that "approve" or "request changes". All users able to submit comment pull request reviews will continue to be able to do so.
```
and there is a checkmark box, which is unchecked, for this option.  This seems to be the correct setting, but perhaps it should be checked?

jkyang92 commented on M2/Macaulay2/e/interface/cone.cpp:
```diff
@@ -0,0 +1,99 @@
+      const size_t c = C->n_cols();  // rank of ambient lattice
+      const size_t r = C->n_rows();  // number of cone inequalities
+
+      auto ineqs = libnormaliz::Matrix<Integer>(r, c);
```
It's slightly better to write ```libnormalize::Matrix<Integer> ineqs(r,c);```. Similarly later for ```libnormalize::Cone<Integer>```. This in theory avoids a copy/move, although here it's almost certainly too cheap to matter.

mahrud replied:
Hmm I think I understand what you mean, but I feel like this is more readable than `ineqs(r, c)` (it almost looks like `ineqs` is a function with two inputs) or `cone(libnormaliz::Type::inequalities, ineqs)`. Do you think the compiler would take care of avoiding this copy/move for us?

jkyang92 replied:
It's pretty standard c++ notation although for c++11 some suggest to always use ```libnormalize::Matrix<Integer> ineqs{r,c};``` note the braces. This avoids the following pitfall: https://en.wikipedia.org/wiki/Most_vexing_parse.

As for avoiding the copy, usually. In c++17, the compiler will always get rid of the copy (the copy doesn't exist in c++17 because of how the standard implements initialization), but prior to that, it's allowed to and usually it does, but it's not technically required.

jkyang92 commented:
@mahrud The memory management is fine. I added a comment on a minor unrelated thing I noticed.

DanGrayson commented:
Could you resolve the merge conflicts?

d-torrance commented:
@mahrud: Should any changes from this PR be mentioned in the 1.24.11 changelog?
